icu4c/source/tools/toolutil/writesrc.cpp

```diff
@@ -357,6 +358,69 @@ usrc_writeUCPTrie(FILE *f, const char *name, const UCPTrie *pTrie, UTargetSyntax
     usrc_writeUCPTrieStruct(f, line, pTrie, line2, line3, line4, syntax);
 }
 
+U_CAPI void U_EXPORT2
+usrc_writeUnicodeSet(
+        FILE *f,
+        const char *name,
+        const USet *pSet,
+        UTargetSyntax syntax) {
+    // ccode is not yet supported
+    U_ASSERT(syntax == UPRV_TARGET_SYNTAX_TOML);
+
+    fprintf(f, "name = \"%s\"\n", name);
+
+    // First write as the uset serialized format
+    icu::ErrorCode localStatus;
+    int32_t length = uset_serialize(pSet, NULL, 0, localStatus);
+    localStatus.reset();
+    std::unique_ptr<uint16_t[]> arr(new uint16_t[length]);
+    uset_serialize(pSet, arr.get(), length, localStatus);
+    U_ASSERT(localStatus.isSuccess());
+    const char* indent = (syntax == UPRV_TARGET_SYNTAX_TOML) ? "  " : "";
+    usrc_writeArray(f, "serialized = [\n  ", arr.get(), 16, length, indent, "\n]\n");
+
+    // Now write it also as a list of ranges
+    USerializedSet serializedSet;
+    uset_getSerializedSet(&serializedSet, arr.get(), length);
+    fprintf(f, "ranges = [\n");
+    for (int32_t i=0; i<uset_getSerializedRangeCount(&serializedSet); i++) {
+        UChar32 start;
+        UChar32 end;
+        uset_getSerializedRange(&serializedSet, i, &start, &end);
+        fprintf(f, "  [0x%x, 0x%x],\n", start, end);
+    }
+    fprintf(f, "]\n");
+}
+
+U_CAPI void U_EXPORT2
+usrc_writeUCPMap(
+        FILE *f,
+        const char *name,
+        const UCPMap *pMap,
+        UProperty uproperty,
+        UTargetSyntax syntax) {
+    // ccode is not yet supported
+    U_ASSERT(syntax == UPRV_TARGET_SYNTAX_TOML);
+
+    fprintf(f, "name = \"%s\"\n", name);
+
+    // Print out list of ranges
+    UChar32 start = 0, end;
+    uint32_t value;
+    fprintf(f, "ranges = [\n");
+    while ((end = ucpmap_getRange(pMap, start, UCPMAP_RANGE_NORMAL, 0, NULL, NULL, &value)) >= 0) {
+        if (uproperty != UCHAR_INVALID_CODE) {
+            const char* name1 = u_getPropertyValueName(uproperty, value, U_LONG_PROPERTY_NAME);
+            const char* name2 = u_getPropertyValueName(uproperty, value, U_SHORT_PROPERTY_NAME);
+            fprintf(f, "  [0x%x, 0x%x, %d, \"%s\", \"%s\"],\n", start, end, value, name1, name2);
```

Comment:
Please print with `%u` so that large values don't show as negative ones.

Optional: Consider printing with `%lu` and casting the value to `(long)`, or look up the standard macros for printf-ing a uint32_t.

Optional: The numeric value and the value name(s) are redundant with each other...

Reply:
did `%lu` in fb441fecde30f7514096b3e9883576703f853c6c, but I didn't add a cast; isn't `fprintf` supposed to consume the inputs as the correct types?

I wanted to print out both the string names and the numeric values to allow clients to choose the one that works better for them.